dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs

```diff
@@ -0,0 +1,309 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System.ComponentModel;
+using System.Linq;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Json.More;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel.AI.ChatCompletion;
+using Microsoft.SemanticKernel.Connectors.AI.OpenAI;
+using Microsoft.SemanticKernel.Connectors.AI.OpenAI.AzureSdk;
+using Microsoft.SemanticKernel.Diagnostics;
+using Microsoft.SemanticKernel.Functions.OpenAPI.Model;
+using Microsoft.SemanticKernel.Orchestration;
+using Microsoft.SemanticKernel.TemplateEngine;
+using Microsoft.SemanticKernel.TemplateEngine.Basic;
+
+#pragma warning disable IDE0130
+// ReSharper disable once CheckNamespace - Using NS of Plan
+namespace Microsoft.SemanticKernel.Planners;
+#pragma warning restore IDE0130
+
+/// <summary>
+/// A planner that uses OpenAI function calling in a stepwise manner to fulfill a user goal or question.
+/// </summary>
+public class FunctionCallingStepwisePlanner
+{
+    /// <summary>
+    /// Initialize a new instance of the <see cref="FunctionCallingStepwisePlanner"/> class.
+    /// </summary>
+    /// <param name="kernel">The semantic kernel instance.</param>
+    /// <param name="config">The planner configuration.</param>
+    public FunctionCallingStepwisePlanner(
+        IKernel kernel,
+        FunctionCallingStepwisePlannerConfig? config = null)
+    {
+        Verify.NotNull(kernel);
+        this._kernel = kernel;
+        this._chatCompletion = kernel.GetService<IChatCompletion>();
+
+        // Initialize prompt renderer
+        this._promptTemplateFactory = new BasicPromptTemplateFactory(this._kernel.LoggerFactory);
+
+        // Set up Config with default values and excluded plugins
+        this.Config = config ?? new();
+        this.Config.ExcludedPlugins.Add(RestrictedPluginName);
+
+        this._initialPlanPrompt = EmbeddedResource.Read("Stepwise.InitialPlanPrompt.txt");
+        this._stepPrompt = EmbeddedResource.Read("Stepwise.StepPrompt.txt");
+
+        // Create context and logger
+        this._logger = this._kernel.LoggerFactory.CreateLogger(this.GetType());
+    }
+
+    /// <summary>
+    /// Execute a plan
+    /// </summary>
+    /// <param name="question">The question to answer</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>Result containing the model's response message and chat history.</returns>
+    /// <exception cref="SKException">If no goal found or failed to retrieve function from kernel.</exception>
+    public async Task<FunctionCallingStepwisePlannerResult> ExecuteAsync(
+        string question,
+        CancellationToken cancellationToken = default)
```

Comment:
This is a very long method and hard to grok without reading in a lot of detail so it will be hard to maintain. Can you consider extracting some methods to make the logic easier to follow.